File: 2015/day2.go
package main

import (
	"fmt"
	"os"
	"strconv"
	"strings"
)

func check(e error) {
	if e != nil {
		panic(e)
	}
}

func min(a int, b int, c int) int {
	if a <= b && a <= c {
		return a
	} else if b <= a && b <= c {
		return b
	} else {
		return c
	}
}

func least(a int, b int, c int) (least int, second int) {
	if a <= b && a <= c {
		least = a
		if c <= b {
			second = c
		} else {
			second = b
		}
	} else if b <= a && b <= c {
		least = b
		if a <= c {
			second = a
		} else {
			second = c
		}
	} else {
		least = c
		if b <= a {
			second = b
		} else {
			second = a
		}
	}
	return
}

func main() {
	dat, err := os.ReadFile("day2.input")
	check(err)

	sdat := strings.Split(string(dat), "\n")
	solve1sum := 0
	solve2sum := 0

	for _, v := range sdat {
		pack := strings.Split(v, "x")
		if v != "" {
			l, _ := strconv.Atoi(pack[0])
			w, _ := strconv.Atoi(pack[1])
			h, _ := strconv.Atoi(pack[2])
			//part1
			slack := min(l*w, w*h, h*l)
			area := 2*l*w + 2*w*h + 2*h*l
			solve1sum += slack + area

			//part2
			least, second := least(l, w, h)
			ribbon := least*2 + second*2
			bow := l * w * h
			solve2sum += ribbon + bow
		}
	}
	fmt.Println(solve1sum)
	fmt.Println(solve2sum)
}

File: 2024/day2.py
#!/usr/bin/env python3


with open("./day2.input", "r") as f:
    data = [[int(y) for y in x.split()] for x in f.read().strip().split("\n")]


part1 = 0
part2 = 0

for i in data:
    inc = set([i[j] - i[j + 1] for j in range(len(i) - 1)])
    if inc <= {1, 2, 3} or inc <= {-1, -2, -3}:
        part1 += 1

    else:
        x = [i[: j - 1] + i[j:] for j in range(1, len(i) + 1)]
        for y in x:
            inc = set([y[j] - y[j + 1] for j in range(len(y) - 1)])
            if inc <= {1, 2, 3} or inc <= {-1, -2, -3}:
                part2 += 1
                break

print(part1)
print(part1 + part2)
